substrate/utils/frame/benchmarking-cli/src/storage/read.rs

```diff
@@ -109,16 +173,71 @@ impl StorageCmd {
 			child_nodes.shuffle(&mut rng);
 
 			info!("Reading {} child keys", child_nodes.len());
+			let (last_child_key, last_child_info) = child_nodes.last().unwrap();
 			for (key, info) in child_nodes.as_slice() {
+				on_validation_batch.push((key.0.clone(), Some(info.clone())));
 				let start = Instant::now();
 				let v = backend
 					.child_storage(info, key.0.as_ref())
 					.expect("Checked above to exist")
 					.ok_or("Value unexpectedly empty")?;
-				record.append(v.len(), start.elapsed())?;
-
+				on_validation_size += v.len();
+				if !self.params.on_block_validation {
+					record.append(v.len(), start.elapsed())?;
+				}
 				read_in_batch += 1;
-				if read_in_batch >= self.params.batch_size {
+				let is_batch_full = read_in_batch >= self.params.batch_size ||
+					(last_child_key == key && last_child_info == info);
+
+				// Read child keys on block validation
+				if is_batch_full && self.params.on_block_validation {
+					let root = backend.root();
```

Comment:
What's under IF repeats two times, maybe we can extract it in a single function.